src/java.base/share/classes/sun/security/pkcs/PKCS8Key.java

```diff
@@ -218,17 +285,45 @@ public String getFormat() {
      */
     private synchronized byte[] getEncodedInternal() {
         if (encodedKey == null) {
-            DerOutputStream tmp = new DerOutputStream();
-            tmp.putInteger(V1);
-            algid.encode(tmp);
-            tmp.putOctetString(key);
-            DerValue out = DerValue.wrap(DerValue.tag_Sequence, tmp);
-            encodedKey = out.toByteArray();
-            out.clear();
+            try {
+                encodedKey = generateEncoding();
+            } catch (IOException e) {
+                e.printStackTrace();
+                // encodedKey is still null
+            }
         }
         return encodedKey;
     }
 
+    private byte[] generateEncoding() throws IOException {
+        DerOutputStream out = new DerOutputStream();
+        out.putInteger((pubKeyEncoded == null) ? 0 : 1);
+        algid.encode(out);
+        out.putOctetString(privKeyMaterial);
+
+        if (attributes != null) {
+            out.writeImplicit(
+                DerValue.createTag((byte) (DerValue.TAG_CONTEXT |
+                    DerValue.TAG_CONSTRUCT), false, (byte) 0),
+                new DerOutputStream().putOctetString(attributes));
+
+        }
+
+        if (pubKeyEncoded != null) {
```

Comment:
Looking back at an earlier conversation between you and Weijun, I think I read that pubKeyEncoded will be set/overwritten if the private key encoding holds a public key.  So when consuming a PKCS#8 EC key, where the private key is itself a SEC1-v2 formatted key encoding with a pubkey, wouldn't the version be set to 0 (v1), but the pubKeyEncoded is also non-null?
I ask only because upon running this method, wouldn't you end up making the output a v2 OneAsymmetricKey, still with the SEC1-v2 private key (with pub key) and also have it in the attributes section?